input-profile.js

```diff
@@ -0,0 +1,249 @@
+                this.controllerObject = this.preSpawnedController;
+                this.controllerObject.parent = this.object;
+                this.controllerObject.setTranslationLocal([0, 0, 0]);
+                this.getGamepadObjectsFromProfile(this.ProfileJSON, this.controllerObject);
+                this.modelLoaded = true;
+                console.log('Loaded ' + this.handedness + '  controller');
+            } else {
+                const assetPath = this.profile + '/' + this.handedness + '.glb';
+                this.engine.scene
+                    .append(assetPath)
+                    .then((obj) => {
+                        this.controllerObject = obj;
+                        this.components = [];
+                        this.getComponents(obj);
+                        this.controllerObject.parent = this.object;
+                        this.controllerObject.setTranslationLocal([0, 0, 0]);
+                        this.getGamepadObjectsFromProfile(
+                            this.ProfileJSON,
+                            this.controllerObject
+                        );
+                        this.modelLoaded = true;
+                        console.log('Loaded ' + this.handedness + '  controller');
+                    })
+                    .catch(console.error);
+            }
+        }
+    }
+
+    getGamepadObjectsFromProfile(profile, obj) {
+        //initialise components
+        const components = profile.layouts[this.handedness].components;
+        if (!components) return;
+        for (const i in components) {
+            if (components.hasOwnProperty(i)) {
+                const visualResponses = components[i].visualResponses;
+                for (const j in visualResponses) {
+                    if (visualResponses.hasOwnProperty(j)) {
+                        this.getObjectByName(obj, visualResponses[j].valueNodeName);
+                        this.getObjectByName(obj, visualResponses[j].minNodeName);
+                        this.getObjectByName(obj, visualResponses[j].maxNodeName);
+                    }
+                }
+            }
+        }
+    }
+
+    getObjectByName(obj, name) {
+        if (!obj || !name) return;
+        if (obj.name === name) this._gamepadObjects[name] = obj;
+        const children = obj.children;
+        for (let i = 0; i < children.length; ++i) this.getObjectByName(children[i], name);
+    }
+
+    assignTransform(target, min, max, value) {
+        vec3.lerp(
+            _TempVec,
+            min.getTranslationWorld([]),
+            max.getTranslationWorld([]),
+            value
+        );
+        target.setTranslationWorld(_TempVec);
+
+        quat.lerp(_TempQuat, min.rotationWorld, max.rotationWorld, value);
+        quat.normalize(_TempQuat, _TempQuat);
+        target.rotationWorld = _TempQuat;
+    }
+
+    mapGamepadInput() {
+        const components = this.ProfileJSON.layouts[this.handedness].components;
```

Comment:
I think it would make more sense to convert the incoming JSON into a more efficient data structure, especially to avoid comparing string constantly